packages/edit-navigation/src/store/actions.js

```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
+		.dispatch( coreDataStore )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: false,
+		} );
+	try {
+		// Ensure all the menu items are available before we start creating placeholders.
+		await registry
+			.resolveSelect( coreDataStore )
+			.getMenuItems( { menus: menuId, per_page: -1 } );
 
-	const mapping = yield getMenuItemToClientIdMapping( post.id );
-	const clientIdToMenuId = invert( mapping );
-
-	const stack = [ post.blocks[ 0 ] ];
-	while ( stack.length ) {
-		const block = stack.pop();
-		if ( ! ( block.clientId in clientIdToMenuId ) ) {
-			const menuItem = yield apiFetch( {
-				path: `/__experimental/menu-items`,
-				method: 'POST',
-				data: {
-					title: 'Placeholder',
-					url: 'Placeholder',
-					menu_order: 1,
-				},
-			} );
+		const menuItemIdToBlockId = await dispatch(
+			getEntityRecordIdToBlockIdMapping( post.id )
+		);
+		const knownBlockIds = new Set( Object.values( menuItemIdToBlockId ) );
 
-			mapping[ menuItem.id ] = block.clientId;
-			const menuItems = yield resolveMenuItems( menuId );
-			yield dispatch(
-				'core',
-				'receiveEntityRecords',
-				'root',
-				'menuItem',
-				[ ...menuItems, menuItem ],
-				menuItemsQuery( menuId ),
-				false
-			);
+		const blocks = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+		for ( const { block } of blocks ) {
+			if ( block.name !== 'core/navigation-link' ) {
+				continue;
+			}
+			if ( ! knownBlockIds.has( block.clientId ) ) {
+				const menuItem = await dispatch(
+					createPlaceholderMenuItem( block, menuId )
+				);
+				menuItemIdToBlockId[ menuItem.id ] = block.clientId;
+			}
 		}
-		stack.push( ...block.innerBlocks );
+
+		dispatch( {
+			type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+			postId: post.id,
+			mapping: menuItemIdToBlockId,
+		} );
+	} finally {
+		registry.dispatch( coreDataStore ).__unstableReleaseStoreLock( lock );
 	}
+};
 
-	yield {
-		type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
-		postId: post.id,
-		mapping,
-	};
-} );
+/**
+ * Creates a single placeholder menu item a specified block without an associated menuItem.
+ * Requests POST /wp/v2/menu-items once for every menu item created.
+ *
+ * @param {Object} block  The block to create a menu item based on.
+ * @param {number} menuId Menu id to embed the placeholder in.
+ * @return {Function} An action creator
+ */
+export const createPlaceholderMenuItem = ( block, menuId ) => async ( {
+	registry,
+} ) => {
+	const existingMenuItems = await registry
+		.select( coreDataStore )
+		.getMenuItems( { menus: menuId, per_page: -1 } );
+
+	const createdMenuItem = await apiFetch( {
+		path: `/__experimental/menu-items`,
+		method: 'POST',
+		data: {
+			title: 'Placeholder',
+			url: 'Placeholder',
+			menu_order: 1,
+		},
+	} );
+
+	await registry
+		.dispatch( coreDataStore )
+		.receiveEntityRecords(
+			'root',
+			'menuItem',
+			[ ...existingMenuItems, createdMenuItem ],
+			menuItemsQuery( menuId ),
+			false
+		);
```

Comment:
The problem here was with having `menuId` in `query`: we don't want to append placeholder to menu yet, but we want this entity to show up on `select()`.

That being said, maybe we don't need the placeholders at all anymore? They are a customizer relict. Let's see if we can remove them.

Reply:
Actually we may be able to do that, we'd just need to send a "preflight" batch to create all the placeholders for us, and then reuse these IDs. Should be simpler than what we're doing here and we will also be able to get rid of locks.

Let's follow up with that change – it will be much simpler without the "client id – block id" mapping thing, and I think we'll have a small debate about using an `__internalMenuItemId` vs `id` attribute.

Reply:
I think there should be three batch requests to the menu item endpoint. 
1. Create all new menu items. Update react state with all new ids. 
2. Update existing menu ids, using new ids and new positions. 
3. Delete any remove menu items. 

The ordering here is important. Deleting last makes sense, as if any the other request fail, then at least you try and restore data. Create items first is needed as you need the ids. 

Breaking into three requests as the other benefit, of making extremely unlike that multiple create/edit/deletes batched requests. Doing it this way is also more restful.

Reply:
All good ideas 👍